src/sparsebase/utils/converter/converter.h

```diff
@@ -13,34 +13,27 @@
 #include <optional>
 #include <tuple>
 #include <unordered_map>
+#include <algorithm>
+#include <cstring>
+#include <fstream>
+#include <functional>
+#include <memory>
+#include <typeindex>
+#include <typeinfo>
+#include <vector>
 
 #include "sparsebase/config.h"
-
-#ifdef USE_CUDA
-#include "sparsebase/format/cuda/format.cuh"
-#include "sparsebase/utils/converter/cuda/converter.cuh"
-#endif
+#include "sparsebase/utils/utils.h"
 
 // Forward declerations for the `Convert` functions in
 // sparsebase/format/format.h
 namespace sparsebase {
-namespace utils {
-namespace converter {
-class Converter;
-template <class ConverterType>
-class ConverterImpl;
-template <typename IDType, typename NNZType, typename ValueType>
-class ConverterOrderTwo;
-template <typename ValueType>
-class ConverterOrderOne;
-
-}  // namespace converter
-}  // namespace utils
-}  // namespace sparsebase
-
-#include "sparsebase/format/format.h"
-namespace sparsebase {
-
+namespace format {
+class Format;
+}
+namespace context {
+class Context;
+}
 namespace utils {
 
 namespace converter {
```

Comment:
So as far as I can see, the Converter class has no constructor defined. So it will have a default public constructor. If this is the case, this is not a true singleton since the users can still construct a Converter themselves. This is not necessarily a bad thing if we have a use case for it but I am not 100% sure.

Reply:
Actually, converters aren't singletons. The ConverterStore is the one that's the singleton. Technically you could create a Converter and register functions to it so I don't think we should keep the constructor protected or anything.

Reply:
So in this case all `Converter` classes would have private constructors. And they all should declare `ConverterStore` a friend class.

Reply:
We are not making it so that `Converter` objects can only be declared through the `ConverterStore`, though. The `ConverterStore` provides the initial converter for every format, but users can create a converter instance in their own code, register functions to that instance, and assign it to the format.

Reply:
True, but the utility of keeping them public and open for users' access I think justifies keeping them public.